src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -1214,16 +1231,37 @@ namespace cryptonote
     return true;
   }
   //-----------------------------------------------------------------------------------------------
-  size_t core::get_block_sync_size(uint64_t height) const
+  size_t core::get_block_sync_size(uint64_t height, const uint64_t average_blocksize_of_biggest_batch) const
   {
-    static const uint64_t quick_height = m_nettype == TESTNET ? 801219 : m_nettype == MAINNET ? 1220516 : 0;
     size_t res = 0;
     if (block_sync_size > 0)
       res = block_sync_size;
-    else if (height >= quick_height)
-      res = BLOCKS_SYNCHRONIZING_DEFAULT_COUNT;
-    else
-      res = BLOCKS_SYNCHRONIZING_DEFAULT_COUNT_PRE_V4;
+    else {
+      size_t number_of_blocks = BLOCKS_MEDIAN_WINDOW;
+      std::vector<uint64_t> last_n_blocks_weights;
+      m_blockchain_storage.get_last_n_blocks_weights(last_n_blocks_weights, number_of_blocks);
+      uint64_t median_weight = epee::misc_utils::median(last_n_blocks_weights);
+      MINFO("Last " << number_of_blocks
+             << " blocks median size is " << median_weight
+             << " bytes and the max average blocksize in the queue is " << average_blocksize_of_biggest_batch << " bytes");
+      uint64_t projected_blocksize = (average_blocksize_of_biggest_batch > median_weight) ? average_blocksize_of_biggest_batch : median_weight;
+      if ((projected_blocksize * BLOCKS_MEDIAN_WINDOW) < batch_max_weight) {
+        res = BLOCKS_MEDIAN_WINDOW;
+        MINFO("blocks are tiny, " << projected_blocksize << " bytes, sync " << res << " blocks in next batch");
+      }
+      else if (projected_blocksize >= batch_max_weight) {
+        res = 1;
+        MINFO("blocks are projected to surpass " << batch_max_weight << " bytes, syncing just a single block in next batch");
+      }
+      else if (projected_blocksize > BLOCKS_HUGE_THRESHOLD_SIZE) {
+        res = 1;
```

Comment:
Why have this separate value, when `batch_max_weight` is already limited?

Reply:
This is an extreme case. If a block is 10mb [1], then we are okay with syncing a batch with single block in it. 

Ordinarily we want to have as many blocks as possible inside a single batch. 


1. To be specific, it is not fixed value of 10mb. It is half of the maximum possible size of the batch. Here the maximum possible size of batch is 20mb. Then if a block is 10mb, then we consider it huge.